contents/kotlin/NullSafety.md

```diff
@@ -0,0 +1,142 @@
+
+
+## Operators in Kotlin
+Although non-nullable type is a strong feature in Kotlin, the [interoperability](https://kotlinlang.org/docs/reference/java-interop.html) with Java means that we have to use variables as nullable type sometimes. In the previous section, we seem to have hit an obstacle as the compiler blocks the call to `secondString.length`. In this section we look at some ways of overcoming this problem.
+### Safe call operator
+Represented by `?.`, the safe call operator is used in this way  
+```kotlin
+secondString?.length
+```
+This returns the length of `secondString` if `secondString` is not `null`, and `null` otherwise.
+
+Now we can chain like this
+```Kotlin
+bob?.department?.manager?.name
+```
+This chain will return `null` if any of the variables inside the chain is `null`.
+
+
+### Elvis Operators
+Represented by `?:`, the Elvis operator is used in this way
+```kotlin
+val length = secondString?.length ?: -1
+```
+If the expression to the left of `?:` is not null, the Elvis operator (`?.`) will return it as it is; else it will return a default value supplied (-1 in this case).
+
+We also notice the use of safe call operator together with Elvis operator in the same statement.
+
+But the Elvis operator is more powerful than this. `return` and `throw` statements are legitimate default values on the right side of the Elvis operator. For example:
+```kotlin
+fun myFunc(node: Node): String? {
+    val parent = node.getParent() ?: return null
+    val name = node.getName() ?: throw IllegalArgumentException("Name expected")
+    // ...
+}
+```
+Doing so like this can help programmers to check for function arguments before carrying on with the required computation.
+
+At this point you may ask, "What if I still want my NPE?"
+
+
+### Not-null assertion operator
+Represented by `!!`, the not-null assertion operator is used in this way
+```kotlin
+val stringLength = secondString!!.length
+```
+The operator converts any value to a non-null type and throws an exception if the value is null. In the above example, `stringLength` will be assigned the length of `secondString` if `secondString` is not `null`; if secondString is `null`, a NPE is thrown. Kotlin tries to reduce the number of NPE as an NPE is difficult to debug a generic NPE, and also it creates so many app crashes. Hence NPEs in Kotlin are explicitly asked for.
+
+
+# Summary
+1. Kotlin makes your applications safer by allocating more work to the complier, instead of failing in the hands of the users.
+2. If you expect your object to **not** take on a `null` value, make it a non-null type!
+3. Even if you make your object a nullable type, Kotlin handles it better than Java because it can help to prevent NPE. An generic NPE is hard to debug; in Kotlin a descriptive message could be given to make debugging easier(with the help of Elvis operator).
```

Comment:
This point about descriptive messages was not mentioned in the main text?

Reply:
It was implied in the section and example on Elvis Operator. I've added a sentence in the Elvis Operator section to make it more obvious.